app/user/user_main.c

```diff
@@ -123,14 +123,27 @@ void nodemcu_init(void)
     }
 
 #if defined(FLASH_SAFE_API)
-    if( flash_safe_get_size_byte() != flash_rom_get_size_byte()) {
-        NODE_ERR("Self adjust flash size.\n");
-        // Fit hardware real flash size.
-        flash_rom_set_size_byte(flash_safe_get_size_byte());
+    uint32_t flash_safe_size_byte = flash_safe_get_size_byte();
+    uint32_t flash_rom_size_byte = flash_rom_get_size_byte();
+    if( flash_safe_size_byte != flash_rom_size_byte ) {
+        if( flash_safe_size_byte <= FLASH_SIZE_4MBYTE ) {
+            NODE_ERR("Self adjust flash size.\n");
+            // Fit hardware real flash size.
+            flash_rom_set_size_byte(flash_safe_size_byte);
+
+            system_restart ();
+            // Don't post the start_lua task, we're about to reboot...
+            return;
+        } else if( flash_rom_size_byte < FLASH_SIZE_1MBYTE) {
+            NODE_ERR("Locking flash size for SDK to 1MByte.\n");
+            // SDK/ROM can't handle flash size > 4MByte, ensure a minimum of 1MByte for firmware image
+            flash_rom_set_size_byte(FLASH_SIZE_1MBYTE);
```

Comment:
Yes, they probably would fail. But as far as I can see, the firmware already established Espressif's example to wrap the flash functions with `flash_safe_read()` and friends that temporarily extend to maximum address range.
At least it works without issues for me ATM.